backend/question-service/lambda/app/scraper/handler.py

```diff
@@ -0,0 +1,189 @@
+import time
+import datetime
+import requests
+from bs4 import BeautifulSoup, XMLParsedAsHTMLWarning
+from pymongo import MongoClient, UpdateOne
+import json
+from pymongo.errors import ConnectionFailure, ServerSelectionTimeoutError, BulkWriteError
+from dotenv import load_dotenv
+import os 
+import warnings
+import concurrent.futures
+
+# Load environment variables
+load_dotenv()
+MONGO_CONNECTION_STRING = os.getenv("MONGO_CONNECTION_STRING")
+RATE_LIMIT = int(os.getenv("RATE_LIMIT", 60)) # Default to 60 questions per function invokation - the Rate Limit
+COOLDOWN_PERIOD = int(os.getenv("COOLDOWN_PERIOD", 10 * 24 * 60 * 60)) # Default to 10 days in seconds
+
+# Set MongoDB Client and Connection as a global variable
+client = None
+collection = None
+
+def handler(event, context):
+    """Lambda function handler. Scrapes Leetcode for questions and saves them to the MongoDB database."""
+
+    # TODO: Implement custom event handling logic
+    print("The event received was: ", event)
+
+    # Establish a connection to the MongoDB server
+    global client
+    client = connect_to_mongo(MONGO_CONNECTION_STRING)
+
+    # Select the database and collection
+    db = client['questions_db']
+    global collection
+    collection = db['problems']
+
+    all_questions = parse_questions_xml()
+    questions_to_scrape = get_questions_to_scrape(all_questions)
+    problems = scrape_leetcode_from_urls(questions_to_scrape)
+    save_to_mongodb(problems)
+
+    # Close the connection
+    client.close()
+
+    return f"Successfully scraped and saved {len(problems)} questions!"
```

Comment:
very nit: for easier debugging in the future might be useful to list the questions scraped

Reply:
Resolved in de1eb1590e1c47555ff72b5033e505e90f480449:
- Added new response object for function response
- Response status indicates whether the function has successfully scraped and saved problems: 'success', 'partial success' (if any questions to scrape failed), 'fail' (if an error occurred during execution)
- Response body now lists questions scraped and questions failed to scrape

An example response is attached below:
<img width="865" alt="image" src="https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g03/assets/48595194/6f37809f-2a45-49fa-af8d-52e300c2cc12">